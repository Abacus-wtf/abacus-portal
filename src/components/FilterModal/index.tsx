import {
  PricingSessionFilters,
  pricingSessionWhere,
} from "@state/sessionData/queries"
import React, { FunctionComponent, FormEvent, useState } from "react"
import styled from "styled-components"
import { Modal, ModalBody, ListGroupItem, ListGroup, Form } from "shards-react"
import { InputWithTitle } from "@components/Input"
import Button from "@components/Button"
import { SessionState } from "@state/sessionData/reducer"
import { theme } from "@config/theme"
import { Label } from "../global.styles"

const StyledModalBody = styled(ModalBody)`
  overflow-y: scroll;
  max-height: 100vh;

  @media ${theme.mediaMin.splitCenter} {
    overflow: unset;
  }
`

const StyledSessionStateContainer = styled.div`
  grid-gap: 10px;
  margin: 10px 0px;
  @media ${theme.mediaMin.splitCenter} {
    display: grid;
    grid-template-columns: 1fr 1fr;
  }
`

type FilterModalProps = {
  open: boolean
  toggle: () => void
  applyFilters: (where: string | null) => Promise<void>
  setFilters: (where: string | null) => void
}

type FilterForm<Elements> = Elements & {
  nftAddress: HTMLInputElement
  tokenId: HTMLInputElement
}

const INPUT_IDS = {
  TOKEN_ID: "tokenId",
  NFT_ADDRESS: "nftAddress",
  Vote: "vote",
  Weigh: "weigh",
  SetFinalAppraisal: "setFinalAppraisal",
  Harvest: "harvest",
  Claim: "claim",
  Complete: "complete",
}

const FilterModal: FunctionComponent<FilterModalProps> = ({
  open,
  toggle,
  applyFilters,
  setFilters,
}) => {
  const [nftAddressValue, setNftAddressValue] = useState("")
  const [tokenIdValue, setTokenIdValue] = useState("")
  const [sessionStatuses, setSessionStatuses] = useState(new Set<number>())

  const toggleSessionState = (state: SessionState) => () =>
    setSessionStatuses((statuses) => {
      if (statuses.has(state)) {
        statuses.delete(state)
      } else {
        statuses.add(state)
      }
      return new Set(statuses)
    })

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    const { nftAddress, tokenId } = e.currentTarget as FilterForm<
      typeof e.currentTarget
    >
    const statuses = Array.from(sessionStatuses)
    const filters: PricingSessionFilters = {
      ...(tokenId.value && { tokenId: Number(tokenId.value) }),
      ...(nftAddress.value && { nftAddress: nftAddress.value }),
      ...(statuses.length && {
        sessionStatus: statuses,
      }),
    }
    const where = pricingSessionWhere(filters)
    applyFilters(where)
    setFilters(where)
    toggle()
  }

  return (
    <Modal size="md" open={open} toggle={toggle} centered>
      <StyledModalBody>
        <h4>Select Filters</h4>
        <Form onSubmit={handleSubmit}>
          <ListGroup>
            <ListGroupItem>
              <InputWithTitle
                title="NFT Address"
                id={INPUT_IDS.NFT_ADDRESS}
                placeholder="0x16baf0de678e52367adc69fd067e5edd1d33e3bf"
                value={nftAddressValue}
                onChange={(e) => setNftAddressValue(e.target.value)}
              />
            </ListGroupItem>
            <ListGroupItem>
              <InputWithTitle
                title="Token ID"
                id={INPUT_IDS.TOKEN_ID}
                placeholder="6407"
                value={tokenIdValue}
                onChange={(e) => setTokenIdValue(e.target.value)}
              />
            </ListGroupItem>
          </ListGroup>
          <Label htmlFor="session-state" style={{ margin: "25px 0 5px 5px" }}>
            Session Status:
          </Label>
          <StyledSessionStateContainer id="session-state">
            <ListGroup>
              <ListGroupItem>
                <InputWithTitle
                  title="Vote"
                  id={INPUT_IDS.Vote}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.Vote)}
                  onChange={toggleSessionState(SessionState.Vote)}
                />
              </ListGroupItem>
              <ListGroupItem>
                <InputWithTitle
                  title="Weigh"
                  id={INPUT_IDS.Weigh}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.Weigh)}
                  onChange={toggleSessionState(SessionState.Weigh)}
                />
              </ListGroupItem>
              <ListGroupItem>
                <InputWithTitle
                  title="Set Final"
                  id={INPUT_IDS.SetFinalAppraisal}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.SetFinalAppraisal)}
                  onChange={toggleSessionState(SessionState.SetFinalAppraisal)}
                />
              </ListGroupItem>
            </ListGroup>
            <ListGroup>
              <ListGroupItem>
                <InputWithTitle
                  title="Harvest"
                  id={INPUT_IDS.Harvest}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.Harvest)}
                  onChange={toggleSessionState(SessionState.Harvest)}
                />
              </ListGroupItem>
              <ListGroupItem>
                <InputWithTitle
                  title="Claim"
                  id={INPUT_IDS.Claim}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.Claim)}
                  onChange={toggleSessionState(SessionState.Claim)}
                />
              </ListGroupItem>
              <ListGroupItem>
                <InputWithTitle
                  title="Complete"
                  id={INPUT_IDS.Complete}
                  type="checkbox"
                  checked={sessionStatuses.has(SessionState.Complete)}
                  onChange={toggleSessionState(SessionState.Complete)}
                />
              </ListGroupItem>
            </ListGroup>
          </StyledSessionStateContainer>

          <Button style={{ width: "100%", marginTop: "15px" }} type="submit">
            Apply Filters
          </Button>
        </Form>
      </StyledModalBody>
    </Modal>
  )
}

export default FilterModal
