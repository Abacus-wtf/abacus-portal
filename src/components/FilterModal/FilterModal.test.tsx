import { render, fireEvent } from "@test-utils/index"
import React from "react"
import "@testing-library/jest-dom/extend-expect"
import FilterModal from "."

const toggle = jest.fn()
const applyFilters = jest.fn()
const setFilters = jest.fn()

const defaultProps = {
  open: true,
  toggle,
  applyFilters,
  setFilters,
}

describe("FilterModal", () => {
  // eslint-disable-next-line jest/no-commented-out-tests
  /* it("Matches snapshot", () => {
    const component = render(<FilterModal {...defaultProps} />)
    expect(component.container).toMatchSnapshot()
  }) */

  it("calls applyFilters and setFilters on submit", () => {
    const { getByText } = render(<FilterModal {...defaultProps} />)

    fireEvent.click(getByText(/Apply Filters/i))
    expect(applyFilters).toHaveBeenCalledTimes(1)
    expect(setFilters).toHaveBeenCalledTimes(1)
  })
})
