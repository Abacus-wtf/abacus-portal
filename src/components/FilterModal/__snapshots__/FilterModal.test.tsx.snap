// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterModal Matches snapshot 1`] = `
<div>
  <div
    class="modal-backdrop fade"
  />
  <div
    aria-hidden="true"
    class="modal fade top"
    role="dialog"
    style="display: block;"
  >
    <div
      class="modal-dialog modal-md modal-dialog-centered"
      role="document"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-body sc-kstqJO bPizxN"
        >
          <h4>
            Select Filters
          </h4>
          <form
            class=""
          >
            <ul
              class="list-group"
            >
              <li
                class="list-group-item"
              >
                <div
                  class="sc-dQoVA QhqRQ"
                  type=""
                >
                  <label
                    class="sc-iqHYmW ibuPqX"
                    style="margin-bottom: 10px;"
                  >
                    NFT Address
                  </label>
                  <input
                    class="sc-crrszt coXBYq form-control form-control-lg"
                    id="nftAddress"
                    inputtype=""
                    placeholder="0x16baf0de678e52367adc69fd067e5edd1d33e3bf"
                    theme="[object Object]"
                    value=""
                  />
                </div>
              </li>
              <li
                class="list-group-item"
              >
                <div
                  class="sc-dQoVA QhqRQ"
                  type=""
                >
                  <label
                    class="sc-iqHYmW ibuPqX"
                    style="margin-bottom: 10px;"
                  >
                    Token ID
                  </label>
                  <input
                    class="sc-crrszt coXBYq form-control form-control-lg"
                    id="tokenId"
                    inputtype=""
                    placeholder="6407"
                    theme="[object Object]"
                    value=""
                  />
                </div>
              </li>
            </ul>
            <label
              class="sc-iqHYmW ibuPqX"
              for="session-state"
              style="margin: 25px 0px 5px 5px;"
            >
              Session Status:
            </label>
            <div
              class="sc-hBEYId hUxTgA"
              id="session-state"
            >
              <ul
                class="list-group"
              >
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Vote
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="vote"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Weigh
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="weigh"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Set Final
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="setFinalAppraisal"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
              </ul>
              <ul
                class="list-group"
              >
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Harvest
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="harvest"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Claim
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="claim"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
                <li
                  class="list-group-item"
                >
                  <div
                    class="sc-dQoVA jImrOz"
                    type="checkbox"
                  >
                    <label
                      class="sc-iqHYmW ibuPqX"
                      style="margin-bottom: 0px;"
                    >
                      Complete
                    </label>
                    <input
                      class="sc-crrszt bcUOwn form-control form-control-lg"
                      id="complete"
                      inputtype="checkbox"
                      theme="[object Object]"
                    />
                  </div>
                </li>
              </ul>
            </div>
            <button
              class="sc-bdfBQB jjXNlU btn btn-primary"
              style="width: 100%; margin-top: 15px;"
              type="submit"
            >
              Apply Filters
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
